### ISSUE TITLE
New Pi-Apps GUI

### ISSUE THREAD

**Issue by Botspot** [Botspot/pi-apps#1578]
Hey everyone, Discord seems to be [down](https://discordstatus.com) at the moment, so we'll chat here.

I think the new Pi-Apps interface is complete, but it's best for several eyes to look at it before I upload changes to nearly every major script.
~~Attached is a zip file of my Pi-Apps directory (I took out the `.git` and `data/status` folders)
[pi-apps.zip](https://github.com/Botspot/pi-apps/files/8209426/pi-apps.zip)
Extract the zip anywhere you want, and give it a spin!~~ 
theofficialgman edit: I suggest cloning from the PR into another folder like the /tmp folder:
```
cd /tmp
rm -rf /tmp/pi-apps
git clone https://github.com/Botspot/pi-apps.git
cd pi-apps
git checkout new-gui
# run the gui
./gui
```
The PR can be found here: https://github.com/Botspot/pi-apps/pull/1580

Feedback appreciated.
This new script design is substantially more complicated than all previous versions of the `gui` script, so there may be bugs in the flow of dialog windows.
In addition to YAD's behavior being changed to a dual-pane layout with integrated list-refreshing, I've also rewritten the xlunch implementation to allow integrated list-refreshing. (avoiding having to close and re-open xlunch when selecting a category)

Pinging a few contributors who might be interested in testing it:
@theofficialgman @cycool29 @Itai-Nelken @ryanfortner @mobilegmYT @CleanMachine1 @Jai-JAP @amirdahan

Edit: a previous version of the zip had DIRECTORY hard-coded to ~/pi-apps. It has since been fixed.

**Comment by Sussy-OS** [comment]
I don't really like the width of the new Pi-Apps GUI, do you think you could make it a little wider?

**Comment by theofficialgman** [comment]
```
Timestamps don't match
original file: /home/pi/pi-apps/apps 1646761906 1646759968.1355262890 /home/pi/pi-apps/apps/Conky/description
/home/pi/pi-apps/data/settings 1646417459 1646761826.8195941340 /home/pi/pi-apps/data/settings/Shuffle App list
/home/pi/pi-apps/data/status 1646530925 1646766668.7320171270 /home/pi/pi-apps/data/status/btop++
/home/pi/pi-apps/etc 1646762308 1646762308.6289543710 /home/pi/pi-apps/etc/preload-daemon
/home/pi/pi-apps/icons/categories 1638374705 1638374705.2131248370 /home/pi/pi-apps/icons/categories/Imported.png
/home/pi/pi-apps/apps 1646761906 1646759968.1355262890 /home/pi/pi-apps/apps/Conky/description
/home/pi/pi-apps/preload 1646762895 
/home/pi/pi-apps/api 1646766313 
/home/pi/pi-apps/data/category-overrides  
new timestamp: /tmp/pi-apps/apps 1646761906 1646759968.0000000000 /tmp/pi-apps/apps/Conky/website
/tmp/pi-apps/data/settings 1646417459 1646761826.0000000000 /tmp/pi-apps/data/settings/Shuffle App list
/tmp/pi-apps/data/status 1646530925 /tmp/pi-apps/etc 1646762308 1646762308.0000000000 /tmp/pi-apps/etc/preload-daemon
/tmp/pi-apps/icons/categories 1638374705 1638374705.0000000000 /tmp/pi-apps/icons/categories/Imported.png
/tmp/pi-apps/apps 1646761906 1646759968.0000000000 /tmp/pi-apps/apps/Conky/website
/tmp/pi-apps/preload 1646771853 
/tmp/pi-apps/api 1646766313 
/tmp/pi-apps/data/category-overrides  
Generating list...
```
timestamping.... guess I will need to look at the zip and see what you changed (fyi, I did not put mine at /pi/pi-apps, I put it in /tmp/pi-apps)

**Comment by theofficialgman** [comment]
also yeah.. it doesn't work because of this error `env: ‘./gui’: No such file or directory` that Sussy-OS already showed

**Comment by theofficialgman** [comment]
![Screenshot from 2022-03-08 19-10-21](https://user-images.githubusercontent.com/28281419/157347318-58135058-4969-445c-b81e-de18fbd4d90a.png)
😆 looks like everything is working nicely (not). its running the pi-apps version from my home directory every time I click on an option regardless of where I put the new one...

**Comment by theofficialgman** [comment]
ah right.. simple fix. you needed to add the directory variable
```
      --dclick-action=true --select-action "env pipe=$pipe geometry2=$geometry2 '${DIRECTORY}/$0'" \
      --button="!${DIRECTORY}/icons/search.png!Search":"env pipe=$pipe geometry2=$geometry2 '${DIRECTORY}/$0' '' '' Search/ Search/" \
      --button="!${DIRECTORY}/icons/options.png":2 \
```

**Comment by theofficialgman** [comment]
oh LOL, yeah never do this botspot
`  xrandr_out="$(xrandr | grep "HDMI-1" | awk '{print $4}' | tr 'x+' ' ')"`
thats guaranteed to pretty much never work

**Comment by Botspot** [comment]
> ah right.. simple fix. you needed to add the directory variable
> 
> ```
>       --dclick-action=true --select-action "env pipe=$pipe geometry2=$geometry2 '${DIRECTORY}/$0'" \
>       --button="!${DIRECTORY}/icons/search.png!Search":"env pipe=$pipe geometry2=$geometry2 '${DIRECTORY}/$0' '' '' Search/ Search/" \
>       --button="!${DIRECTORY}/icons/options.png":2 \
> ```

Strange. If being run from an absolute path, the $0 variable should also be an absolute path.

**Comment by theofficialgman** [comment]
getting the output from Screen 0 is safe and "should" work everywhere
some example outputs:
```
Screen 0: minimum 8 x 8, current 1920 x 1080, maximum 16384 x 16384
DSI-0 connected (normal left inverted right x axis y axis)
   720x1280      60.00 +
DP-0 connected primary 1920x1080+0+0 (normal left inverted right x axis y axis) 480mm x 270mm
   1920x1080     60.00*+  59.95    50.00  
   1280x1024     75.03    60.00  
   1280x720      60.00    59.94    50.00  
   1152x864      75.00  
   1024x768      75.03    60.01  
   800x600       75.00    60.32  
   720x576       50.00  
   720x480       59.94  
   720x400       70.04  
   640x480       75.00    59.94    59.94
   ```
 ```
 Screen 0: minimum 8 x 8, current 1280 x 720, maximum 16384 x 16384
DSI-0 connected primary 1280x720+0+0 left (normal left inverted right x axis y axis) 0mm x 0mm
   720x1280      60.00*+
DP-0 disconnected (normal left inverted right x axis y axis)
```

**Comment by theofficialgman** [comment]
> > ah right.. simple fix. you needed to add the directory variable
> > ```
> >       --dclick-action=true --select-action "env pipe=$pipe geometry2=$geometry2 '${DIRECTORY}/$0'" \
> >       --button="!${DIRECTORY}/icons/search.png!Search":"env pipe=$pipe geometry2=$geometry2 '${DIRECTORY}/$0' '' '' Search/ Search/" \
> >       --button="!${DIRECTORY}/icons/options.png":2 \
> > ```
> 
> Strange. If being run from an absolute path, the $0 variable should also be an absolute path.

yeah we aren't running from an absolute path, launched it with `./gui` from the folder as was allowed before

**Comment by Botspot** [comment]
> yeah we aren't running from an absolute path, launched it with `./gui` from the folder as was allowed before

Should be easy to fix then. Every mention of `"$0"` should be changed to `"$DIRECTORY/gui"` and then it should work.

**Comment by theofficialgman** [comment]
@Botspot this should be suitable
```
  #determine screen_width and screen_height
  screen_dimensions="$(xdpyinfo | grep "dimensions" | awk '{ print $2 }' | tr 'x+' ' ')"
  screen_width="$(awk '{print $1}' <<<"$screen_dimensions")"
  screen_height="$(awk '{print $2}' <<<"$screen_dimensions")"
  unset screen_dimensions
  ```

**Comment by Botspot** [comment]
> @Botspot this should be suitable
> ```
>   screen_dimensions="$(xdpyinfo | grep "dimensions" | awk '{ print $2 }' | tr 'x+' ' ')"
> ```

The problem with using `xpdyinfo` is that people with dual displays encountered problems.
See: https://github.com/Botspot/pi-apps/issues/172

**Comment by cycool29** [comment]
> [pi-apps.zip](https://github.com/Botspot/pi-apps/files/8209426/pi-apps.zip)

Nice! 
BTW the app info window should be closed when opening settings.

**Comment by Botspot** [comment]
> Nice! BTW the app info window should be closed when opening settings.

Could you make a screen recording? In my desktop environment, the app Details windows automatically closes when it is no longer focused.

**Comment by theofficialgman** [comment]
> > @Botspot this should be suitable
> > ```
> >   screen_dimensions="$(xdpyinfo | grep "dimensions" | awk '{ print $2 }' | tr 'x+' ' ')"
> > ```
> 
> The problem with using `xpdyinfo` is that people with dual displays encountered problems. See: #172

ok then use xrandr but grep this output from it: `Screen 0: minimum 8 x 8, current 1920 x 1080, maximum 16384 x 16384`
works perfectly fine with dual screens

**Comment by theofficialgman** [comment]
@Botspot this should be fine, we don't really care output the garbage output after the resolutions:

```
  screen_dimensions="$(xrandr | grep "Screen 0:" | sed -n -e 's/^.*current //p' | tr 'x+' ' ' | tr ',+' ' ')"
  screen_width="$(awk '{print $1}' <<<"$screen_dimensions")"
  screen_height="$(awk '{print $2}' <<<"$screen_dimensions")"
  unset screen_dimensions
  ```

**Comment by theofficialgman** [comment]
hmm... actually heck this also has the same issues.

**Comment by theofficialgman** [comment]
going to take the code from neofetch here and massage it a bit

https://github.com/dylanaraps/neofetch/blob/ccd5d9f52609bbdcd5d8fa78c4fdb0f12954125f/neofetch#L3082-L3115
```
  screen_dimensions="$(xrandr --nograb --current |  awk -F 'connected |\\+|\\('  '/ connected.*[0-9]+x[0-9]+\+/ && $2 {printf $2 ", "}' | sed -n -e 's/^.*primary //p' | tr 'x+' ' ' | tr ',+' ' ')"
  screen_width="$(awk '{print $1}' <<<"$screen_dimensions")"
  screen_height="$(awk '{print $2}' <<<"$screen_dimensions")"
  unset screen_dimensions
  ```
  
  this "should" work on both single and multi-monitors, with the primary monitor resolution returned. yes I have tested on a multi-monitor setup without issues

**Comment by theofficialgman** [comment]
another issue to report, right clicking in the popup windows (or clicking in the title bar) causes it to close
these are actions that should be allowed (for example right clicking so a user can copy links out of an apps description)

**Comment by Botspot** [comment]
> another issue to report, right clicking in the popup windows (or clicking in the title bar) causes it to close these are actions that should be allowed (for example right clicking so a user can copy links out of an apps description)

I'm not sure how much can be done, as this is using yad's `--close-on-unfocus` flag.

> do you think you could open a PR for this? it would be easier to track. just make another branch for it in the repo here and you can open a PR from it to master. I can open it for you if you want

Fork added here: https://github.com/Botspot/pi-apps/tree/new-gui

**Comment by theofficialgman** [comment]
> New branch has been added: https://github.com/Botspot/pi-apps/tree/new-gui Does it really make sense to open a PR at this time?

yes, draft PRs are a thing and make it much easier to track issues (we can comment on changes desired and then those can be automatically included by clicking the add commit button)

**Comment by Sussy-OS** [comment]
Hi, @Botspot I've been tracking this issue on my PW 4, and I just now got on my pi. I was wondering, should we mainly talk on this issue, or on @theofficialgman 's draft pull?

**Comment by Botspot** [comment]
> Hi, @Botspot I've been tracking this issue on my PW 4, and I just now got on my pi. I was wondering, should we mainly talk on this issue, or on @theofficialgman 's draft pull?

It doesn't really matter. I guess discussion about particular code bits would make more sense on the pull, and general comments/suggestions can be made either here or there.

**Comment by Sussy-OS** [comment]
> 
> It doesn't really matter. I guess discussion about particular code bits would make more sense on the pull, and general comments/suggestions can be made either here or there.

Okay, that makes sence. Thanks.

**Comment by Crilum** [comment]
I'm not a huge fan of the new GUI, it's just too narrow, and I like being able to install/uninstall apps without opening the description. :) I do kind of like the side by side thing though.

The app list also doesn't work at all on my system (RPiOS 11) :

https://user-images.githubusercontent.com/91354257/157560960-21a30a58-48e3-410e-8343-21816eeabb85.mp4

Search works though, and you can view apps from there..

Also, I don't think the `Multi-Install` button is in the right place, it just seems kind of weird in the Settings.

**Comment by theofficialgman** [comment]
> The app list also doesn't work at all on my system (RPiOS 11) :

@Crilum
thats also the error I was getting before I commited the change about an hour ago
can you do a git pull and make sure its not gone?

**Comment by Crilum** [comment]
@theofficialgman git says everything is up to date.

**Comment by theofficialgman** [comment]
ok... yeah there are actually quite a few bugs
if I keep the color set to nothing, it just overwrites the text color as I move my mouse over it with whatever color happens to actually be set (so if there is green text and if I hover over it, then move over other text that changes as well)

**Comment by theofficialgman** [comment]
@Crilum pull again now. I enabled some things for testing

**Comment by Crilum** [comment]
@theofficialgman it works now

**Comment by theofficialgman** [comment]
ok so I see why botspot changed it to $3 then...
there are differences between your echo command and mine
yours isn't printing a newline on the empty character (but it should be) mine is

**Comment by Crilum** [comment]
Hmm. I've always found echo to be simple, but not very smart... lol

**Comment by theofficialgman** [comment]
@Crilum ok pull and again then try.
if it doesn't work, I want you to the send your `LIST-` file here (in data/preload)

**Comment by Crilum** [comment]
What doesn't work? The app list works now, after you said:
>@Crilum pull again now. I enabled some things for testing

**Comment by theofficialgman** [comment]
> What doesn't work? The app list works now, after you said:
> 
> > @Crilum pull again now. I enabled some things for testing

we can't do what I did for that pull... I manually set the color of text to white which "fixed" it (but that won't work for light themes).
just try again now

**Comment by theofficialgman** [comment]
@Crilum ok I so force pushed some changes. delete your testing repo and re-clone it and try again please.
if it doesn't work, please send the contents of your data/preload folder here

**Comment by theofficialgman** [comment]
for reference this is what LIST- should generate like:
```

/tmp/pi-apps/icons/categories/All Apps.png
All Apps
All Apps/
App folder

/tmp/pi-apps/icons/categories/Appearance.png
Appearance
Appearance/
App folder

/tmp/pi-apps/icons/categories/Editors.png
Editors
Editors/
App folder
```
(spaces are blank color definitons)
and this is LIST-All Apps
```

/tmp/pi-apps/icons/back.png
Back
/
Return to the previous location

/tmp/pi-apps/apps/Alacritty Terminal/icon-24.png
Alacritty Terminal
All Apps/Alacritty Terminal
A fast, cross-platform, OpenGL terminal emulator

/tmp/pi-apps/apps/All Is Well/icon-24.png
All Is Well
All Apps/All Is Well
ALL IS WELL is a bash script that allows users to easily update and upgrade their repositories and packages on linux.
```

anything different will break the format and you get issues...

**Comment by theofficialgman** [comment]
alright... so turns out is its not echo thats broken... its redirection. hence why some OSs get this as output:
```
1
2
3 Appearance
4 Appearance/
```
and others get this:
```
1
2
3 Appearance/
4 App folder
```

**Comment by theofficialgman** [comment]
ok... so here is the bug
this is what pi-apps yad looks like:
`--column=@fore@:hd --column=:IMG --column=Name --column=Sysname:HD --column=tip:HD`

according to yad:
```Special column names @fore@, @back@ and @font@ sets corresponding rows attributes. Values of those columns don't show in results.```
so the color should be hidden from the output, right? NOPE
```
1 #00AA00
2 
```
gosh yad.... fix your docs. even worse there is a bug on top of this

newer yad is combining two null outputs into one.... the .png evaluates to a null output (yeah it outputs null no matter what the png location was), and if the color is not passed in, also evaluates to a null output. this is why if we set the manually set the color everything starts working, because it doesn't remove all null outputs, it just combines them into one.

older yad does not have this bug, it outputs the same number of arguments no matter if they are null or not.

the workaround you ask? either 
A: always set the color (which is probably necessary already since not setting it has bugs already described by me here: https://github.com/Botspot/pi-apps/issues/1578#issuecomment-1063529483)

or
B:change the order of the columns and the preload lists so there are never two null outputs next to each other

**Comment by theofficialgman** [comment]
I've implemented option B: but as I said.. at least on my version of yad, there is a bug where names change colors as the mouse drags over them

https://user-images.githubusercontent.com/28281419/157590653-fb87e7bc-a334-4367-9a24-6fa689aababd.mp4

**Comment by theofficialgman** [comment]
now, in order to fix this bug we need to set the color. which means we need to get what the default color "should" be

this can be accomplished with a simple python script (it doesn't appear there are any gtk bindings for bash)
```
#!/usr/bin/env python
import gi
gi.require_version('Gtk', '3.0')
from gi.repository import Gtk
win = Gtk.Window()
vb = Gtk.VBox()
tv = Gtk.TextView()
vb.add(tv)
win.show_all()
style = tv.get_style_context()
textcolor = style.get_color(Gtk.StateType.NORMAL) 
print(textcolor)
```

for my theme (a dark one) it returns pure white `Gdk.RGBA(red=1.000000, green=1.000000, blue=1.000000, alpha=1.000000)`
I've tested setting a lite theme as well and it returns darker text `Gdk.RGBA(red=0.082353, green=0.086275, blue=0.109804, alpha=0.850000)`

**Comment by theofficialgman** [comment]
and OFC, as is yad fashion, the documentation is wrong again
```
--mode=MODE
Set output color mode. Possible values are hex or rgb. Default is hex. HEX mode looks like #rrggbbaa, RGB mode - rgba(r, g, b, a). In RGBA mode opacity have values from 0.0 to 1.0.
```
nope, RGBA has to be defined from 0-255 for RGB and 0-1 for A (thats normal, but the docs are wrong)

**Comment by theofficialgman** [comment]
alright fixed with https://github.com/Botspot/pi-apps/pull/1580/commits/21890b3afde73b2d3b358ac097c7e9eaef6956d0

if anyone has a better way of getting from pure bash, let me know. but I think python or compiled C/C++ code are the only options here

**Comment by theofficialgman** [comment]
> > if anyone has a better way of getting from pure bash, let me know. but I think python or compiled C/C++ code are the only options here
> 
> A while back, I made a light/dark theme detection script with some help from @cycool29. After some further modifications, here it is:

I appreciate the response botspot, but that won't work at all. for one, it requires LXDE (aka the lxsession folder) which is definitly not a given. for two, your script won't detect if the user has set the GTK theme using the export option (like you can set via pi-apps settings.

the python script has neither of these issues and should be safe as all systems likely have python (2 or 3, it doesn't matter), installed by default with their OS. Worst case you don't have it and the color is set to nothing like before

**Comment by Botspot** [comment]
> I appreciate the response botspot, but that won't work at all. for one, it requires LXDE (aka the lxsession folder) which is definitly not a given.

Not quite. The bash-based detection script also checks for settings in `$HOME/.config/gtk-3.0/settings.ini` and `$HOME/.gtkrc-2.0`, so lxsession is not required.

> for two, your script won't detect if the user has set the GTK theme using the export option (like you can set via pi-apps settings

That would be very easy to fix.

**Comment by theofficialgman** [comment]
> Not quite. The bash-based detection script also checks for settings in `$HOME/.config/gtk-3.0/settings.ini` and `$HOME/.gtkrc-2.0`, so lxsession is not required.

I have neither of these alternative files.... so your bash script fails.
the python script however has no issues

**Comment by Botspot** [comment]
Let's take a step back: we're trying to avoid triggering a bunch of harmless GTK warnings when no color is given in yad. But this only occurs with outdated yad in an outdated OS.
Stretch, Buster, and Bullseye users won't encounter this problem. Is it worth fixing? Perhaps it would be easier to update yad on older OS'es like yours.

**Comment by theofficialgman** [comment]
> Let's take a step back: we're trying to avoid triggering a bunch of harmless GTK warnings when no color is given in yad. But this only occurs with outdated yad in an outdated OS. Stretch, Buster, and Bullseye users won't encounter this problem. Is it worth fixing? Perhaps it would be easier to update yad on older OS'es like yours.

its not the warning that is the problem. its this that is the real issue:
`https://user-images.githubusercontent.com/28281419/157590653-fb87e7bc-a334-4367-9a24-6fa689aababd.mp4`

**Comment by Botspot** [comment]
> > Let's take a step back: we're trying to avoid triggering a bunch of harmless GTK warnings when no color is given in yad. But this only occurs with outdated yad in an outdated OS. Stretch, Buster, and Bullseye users won't encounter this problem. Is it worth fixing? Perhaps it would be easier to update yad on older OS'es like yours.
> 
> its not the warning that is the problem. its this that is the real issue: `https://user-images.githubusercontent.com/28281419/157590653-fb87e7bc-a334-4367-9a24-6fa689aababd.mp4`

I saw the video.
Again, that doesn't occur on yad 0.40.0, so therefore it won't occur in the past three debian versions. I'm not sure how many people use Pi-Apps on Ubuntu 18, but probably not that many.
This python thing could be considered a "hacky fix", so perhaps it would be better to implement an alternative "hacky fix" that simply updates yad to 0.40.0?

**Comment by theofficialgman** [comment]
setting the color isn't really a hack... it should be the default
not setting it lets GTK/yad do whatever it wants with the color. in this case it adopts whatever the cursor was hovered over.

I'm unaware if this is an X11, GTK, or yad change, but I'm sure I wouldn't be the only person to experience it

**Comment by Botspot** [comment]
> Stretch has yad 0.38, fyi. so not the past three. still don't know where the issue comes from though

Could you try installing yad 0.40.0 on your system to see if the problem goes away?

**Comment by theofficialgman** [comment]
I'm leaning on it being a GTK bug (or not a bug, more like a functionality change), I've reviewed the yad commit history and see no mention of this issue. I'll compile yad 0.40.0 once I get home later today and check

**Comment by theofficialgman** [comment]
> I'm leaning on it being a GTK bug (or not a bug, more like a functionality change), I've reviewed the yad commit history and see no mention of this issue. I'll compile yad 0.40.0 once I get home later today and check though

@Botspot bug was in YAD and fixed in 0.40.0
I can reproduce the bug in 0.39.0 and older when I build and test them.
gtk warning spam also goes away in 0.40.0 with no color specified

still looking through the commits to see what actually fixed it... nothing references it. it might have been an accidental fix 😆 

I'm ok with this though: 

> Alright, the color-detection workaround can be used if yad is detected to be a version older than 0.40.0.

**Comment by theofficialgman** [comment]
yad doesn't have svn repo source code for anything after 0.39.0... they stopped uploading it and just make a release tarball on 0.40.0. the issue was not fixed in the latest source code available in the snv repo... so we have no clue what exactly fixed it unless you diff the tarballs

**Comment by theofficialgman** [comment]
ok, implemented: https://github.com/Botspot/pi-apps/pull/1580/commits/27a9f67f9d707647534e12de235eaf97b2079573
should be fine for everyone now

**Comment by theofficialgman** [comment]
@Botspot now thats all out of the way and everything seems stable lets talk plans for this gui.

I added back in the announcements (which now only loads once per session since that can't be re-loaded I think).
Do you have plans for anything else regarding this? I personally think it all looks really clean. 
Maybe the only thing I would think of adding would be some sort of scaling method (increase text fontsize everywhere and windows sizes) for high resolution displays.

Here are a couple of things I really like about this update that haven't been mentioned yet
1. the settings button being in the main window is very nice (I'm sure people missed it entirely being in a separate .desktop file before)
2. website and credits being in links in the app window is nice as opposed to a credits button at the bottom

My main concern with this update:
The UI stays open when installing/uninstalling a script, allowing the user to select more things and get multiple install instances going at once, potentially causing errors.

**Comment by Botspot** [comment]
> @Botspot now thats all out of the way and everything seems stable lets talk plans for this gui.
> 
> I added back in the announcements (which now only loads once per session since that can't be re-loaded I think). Do you have plans for anything else regarding this? I personally think it all looks really clean. Maybe the only thing I would think of adding would be some sort of scaling method (increase text fontsize everywhere and windows sizes) for high resolution displays.
> 
> Here are a couple of things I really like about this update that haven't been mentioned yet
> 
> 1. the settings button being in the main window is very nice (I'm sure people missed it entirely being in a separate .desktop file before)
> 2. website and credits being in links in the app window is nice as opposed to a credits button at the bottom



> My main concern with this update: The UI stays open when installing/uninstalling a script, allowing the user to select more things and get multiple install instances going at once, potentially causing errors.

Some people already seem to be doing that and we've mostly mitigated it with apt_lock_wait.
I think that allowing the UI to be navigable whilst an app is installing is more important now than ever before - installing one app at a time, waiting for the current one to finish before starting the next one, has got to be frustrating.

To avoid multiple instances of `manage` running, I could implement a lockfile, but that would have its own issues. Alternatively, the `manage` script could be changed to use a request/reply architecture: each secondary instance of `manage` would notify a daemon to add the action to a queue. Then, yad could be used to visualize the queue, roughly similar to how gparted's UI works.
What do you think, @theofficialgman?

**Comment by theofficialgman** [comment]
> Some people already seem to be doing that and we've mostly mitigated it with apt_lock_wait. I think that allowing the UI to be navigable whilst an app is installing is more important now than ever before - installing one app at a time, waiting for the current one to finish before starting the next one, has got to be frustrating.

I'm mostly concerned about what the individual app install scripts might be doing at the same time. keeping apt update/install from running at the same time is a must, and as you said thats supposed to be handled by `apt_lock_wait`. I think the problem there is though you expect to be able to access your dummy deb folder for the whole time install_packages is running, not just when apt itself is actually running. 

What could happen is two install_packages get called at similar times. The first one runs the apt update white the other waits in apt_lock_wait. The first one finishes and then goes to create the dummy deb, at this time the second script is down running apt update. First script is done now with creating the dummy deb and goes to install it, second script starts creating its own dummy deb. The first script is trying to install its dummy deb and its Packages file is gone because the second script started making its own dummy deb, overwriting the first scripts one it has already created.

this is what is probably happening when the `/tmp/pi-apps-local-packages/Packages` disappears as seen in some user logs. In order to mitigate this, a few things need to happen. #1, rename the `/tmp/pi-apps-local-packages` folder to use a unique name per script (this will prevent other installs overwriting the folder of another). #2, well now that I've fleshed out #1 I can't remember what 2 was supposed to be... maybe it will come back to me

I am also concerned about what the individual scripts might be doing that might conflict with one another.... stuff like scripts that compile code, if the user happens to pick two script that need a good amount of ram, one of them may fail due to out of ram issues caused by the other one. Error logs then might come in for the failed script which only failed because another script was using all the resources... so we would for sure need to add some "current parallel installs" line to the error logging so we can see what all is happening at once.

> To avoid multiple instances of `manage` running, I could implement a lockfile, but that would have its own issues. Alternatively, the `manage` script could be changed to use a request/reply architecture: each secondary instance of `manage` would notify a daemon to add the action to a queue. Then, yad could be used to visualize the queue, roughly similar to how gparted's UI works. What do you think, @theofficialgman?

well if you only allow one manage instance, that pretty much means only one install can actually be running at once, its just you let the user queue up installs to happen sequentially.  The issue I see there just has to do with how pi-apps is structured. sudo permissions don't pass through between subscripts, so each script always asks for the sudo password the first time its called, no matter if you are within the sudo timeout of another subscript. Users might queue up installs, walk away, then come back to find only the first app installed and the second running script and queue is waiting for a sudo password. (there are solutions for this though if you think you want this sequential route)

**Comment by Crilum** [comment]
> @Crilum ok I so force pushed some changes. delete your testing repo and re-clone it and try again please. if it doesn't work, please send the contents of your data/preload folder here

It works now.

**Comment by Crilum** [comment]
I clicked the update button in the new GUI, and after updating, this happened:

https://user-images.githubusercontent.com/91354257/158033929-b5629fe3-b01d-4f57-8365-c11f7c61aaf2.mp4

**Comment by theofficialgman** [comment]
> I clicked the update button in the new GUI, and after updating, this happened:
> 
>  brokenpi-apps-2022-03-12_13.22.24.mp4

I already told another user this.... don't do that. its expected to break and not my issue.
read here: https://github.com/Botspot/pi-apps/pull/1580#issuecomment-1065927445

**Comment by Botspot** [comment]
> I clicked the update button in the new GUI, and after updating, this happened:

You just "updated" back to the live version of Pi-Apps. The GUI is different, so the app-list will look weird until you close Pi-Apps and launch it again.

**Comment by Crilum** [comment]
ohh.. oops. Ok.

**Comment by Botspot** [comment]
> > Nice! BTW the app info window should be closed when opening settings.
> 
> Could you make a screen recording? In my desktop environment, the app Details windows automatically closes when it is no longer focused.

@cycool29, was this resolved?

**Comment by cycool29** [comment]
> @cycool29, was this resolved?

Not really. It works when I manually focus and unfocus the window, but it did not automatically close when the settings window is launched. Here is the recording:

https://user-images.githubusercontent.com/88134003/159405303-2de8100e-46ea-46a8-a2f0-e962a98c16a9.mp4

**Comment by theofficialgman** [comment]
> > @cycool29, was this resolved?
> 
> Not really. It works when I manually focus and unfocus the window, but it did not automatically close when the settings window is launched. Here is the recording:
> 
>  simplescreenrecorder-2022-03-22_11.57.45.mp4

dang why is everything sooooooo sluggish in your video? is it just because you are recording or do you have some craptastic SD card?

**Comment by cycool29** [comment]
> dang why is everything sooooooo sluggish in your video? is it just because you are recording or do you have some craptastic SD card?

The video plays well for me. Not sure why it is sluggish at your side.


> also the issue is likely just your DE, it doesn't unfocus the window when you click off of it

Then this should be fixed right? I am on a very new Raspberry Pi OS bullseye with VNC enabled (so the window manager changed to openbox).

**Comment by theofficialgman** [comment]
> The video plays well for me. Not sure why it is sluggish at your side.

no not the video itself. your pi-apps is running quite slow to do anything. those lists take a while to generate on your install and it takes amiberry over 7 seconds to open.

on my install it takes lest than 2 second to open the all apps page for the first time (takes you 5) and amiberry loads in 1.5 seconds (takes you 7 seconds). I just think either your cpu was pegged by filming this, or your sd card is really bad

**Comment by cycool29** [comment]
I think I found the problem. 

The app info window needs to be focused before unfocusing so the automatic-close will be triggered.

But when an app is selected from the app list, the app info window is *not* automatically focused, and if we don't manually close or focus the app info window before clicking settings, the app info will remain open.

Is there a way to force close the app info window whenever the settings button is clicked? Or maybe automatically focus on the app info window when an app is chosen from the list?

**Comment by cycool29** [comment]
> no not the video itself. your pi-apps is running quite slow to do anything. those lists take a while to generate on your install and it takes amiberry over 7 seconds to open.
>
> on my install it takes lest than 2 seconds to open the all apps page for the first time (takes you 5) and amiberry loads in 1.5 seconds (takes you 7 seconds). I just think either your cpu was pegged by filming this, or your sd card is really bad

I have a Google Meet tab opened, maybe that's the cause.
I will test again when my meeting ends.

**Comment by Botspot** [comment]
The details window should take focus from the app list. I've never seen it *not* do this.
I'm using Mutter window manager.
https://user-images.githubusercontent.com/54716352/159527659-40483ed6-e6e0-494f-99d6-f902b81c37e7.mp4

**Comment by cycool29** [comment]
I found a way to make it focus on the app info window in openbox, see https://github.com/Botspot/pi-apps/pull/1580#pullrequestreview-918047824.

**Comment by cycool29** [comment]
> no not the video itself. your pi-apps is running quite slow to do anything. those lists take a while to generate on your install and it takes amiberry over 7 seconds to open.
>
> on my install it takes lest than 2 second to open the all apps page for the first time (takes you 5) and amiberry loads in 1.5 seconds (takes you 7 seconds). I just think either your cpu was pegged by filming this, or your sd card is really bad

After my meeting ends pi-apps runs smoothly.

**Comment by Botspot** [comment]
@cycool29, could you try adding the `--no-focus` flag to line 577 in your `/tmp/pi-apps/gui` script and see if that solves the window focusing issue?
